### DISCUSSION TITLE
Retrieving list of currently scheduled workflows

### DISCUSSION THREAD

**Discussion by johnwc** [elsa-workflows/elsa-core#4231] [Design & tech]
@sfmskywalker  Internally to Elsa, is there a store that allows to call a method that returns a list of scheduled workflows, with the next expected execution time being one of the properties?

**Comment by sfmskywalker** [comment]
Yes. Although a workflow instance itself doesn’t know anything about scheduled execution times, the timer activities create bookmarks that do contain scheduled execution times (although these are not stored in columns).

What you could do, therefore, is query all bookmarks of a given activity type (e.g. Timer or Delay or Cron), each of these bookmarks have a serialized state of their payload, which includes the scheduled time stamp. 

If you did need to be able to query by scheduled time, then you’d have to create a separate database table and basically shadow-copy the information you need into that when bookmarks are being indexed.

**Reply by johnwc** [reply]
What I am looking to do is to extend the workflow definition list to include a column titled "Scheduled". This column will display the next start execution time for the workflow, if it has one. What objects/methods do I need to call to get the list of scheduled workflows?

**Reply by sfmskywalker** [reply]
To achieve the desired functionality of displaying the next start execution time for each workflow in a "Scheduled" column, consider the following steps:

1. **Query Bookmarks**: Begin by querying all bookmarks associated with timer activities, such as Timer, Delay, and Cron. These bookmarks contain serialized states of their payloads, which include the scheduled timestamps.

2. **Deserialize Payloads**: Deserialize the payload for each bookmark to extract the scheduled timestamp. To understand the exact types and structure, I recommend consulting the source code of the Timer, Delay, and Cron activities. This will give you insights into the bookmark payload types that are created.

3. **Optimization (Optional)**: If you find that querying and deserializing bookmarks frequently impacts performance, you might want to create a separate database table dedicated to storing the scheduled timestamps. By doing so, every time a new timer activity is scheduled, you can update this table with the pertinent information. This approach will facilitate faster and more efficient querying of scheduled timestamps.
